### PR TITLE
VC-402: Livestream integration with Agora. Basic.

### PR DESCRIPTION
**Pull request by vladimircores**
This is basic working version of Livestream feature. No branded design for UI, only simple controls. This can be already used and open a way to implement posts about livestream in other parts of the application.

┆Issue is synchronized with this [Jira Task](https://highvibe.atlassian.net/browse/HVN1-271) by [Unito](https://www.unito.io/learn-more)

### CODE REVIEW

**Comment by vladimircores** [comment]
This is how it works:
agora_video_live

**Comment by vladimircores** [comment]
And this is how it works:
agora_video_live_10fps-min

File: mobile/lib/modules/live/model/store/LiveStore.dart [hunk 1]
```diff
@@ -0,0 +1,154 @@
+import 'package:agora_rtc_engine/rtc_engine.dart';
```
Comment (AlexVegner) [inline mobile/lib/modules/live/model/store/LiveStore.dart:1]: By Dart convention file and folder names should be in snake case.
And please use _controller for mobex stores 
So please move it to core/lib/controllers/live/live_controller.dart

File: mobile/lib/modules/live/model/store/LiveStore.dart [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+import 'live/AgoraStore.dart';
+import 'live/MasterStore.dart';
+import 'live/enum/LiveStatus.dart';
+import 'live/PeersStore.dart';
+
+part 'LiveStore.g.dart';
+
+class LiveStore = _LiveStore with _$LiveStore;
+
+abstract class _LiveStore with Store {
```
Comment (AlexVegner) [inline mobile/lib/modules/live/model/store/LiveStore.dart:22]: Rename it ad LiveController

File: mobile/lib/modules/live/model/store/LiveStore.dart [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+
+  DocumentReference _roomDocument;
+
+  ClientStore _clientStore;
```
Comment (AlexVegner) [inline mobile/lib/modules/live/model/store/LiveStore.dart:26]: Please inject all dependencies. and keep them final by default.

File: mobile/lib/modules/live/model/store/LiveStore.dart [hunk 4]
```diff
@@ -0,0 +1,154 @@
...
+    if (isRoomClosed) return dispose(false);
+    _onLiveStatus(LiveStatus.PEERS_COLLECTION_READY);
+
+    await _agoraStore.initialize(this.roomVO, size);
+    if (isRoomClosed) return dispose(false);
+    _onLiveStatus(LiveStatus.AGORA_ENGINE_READY);
+
+    if (isMaster)
+    {
+      _masterStore = MasterStore();
```
Comment (AlexVegner) [inline mobile/lib/modules/live/model/store/LiveStore.dart:105]: Please inject it, to be sure that we could write unit test in future.

File: mobile/lib/modules/live/model/store/LiveStore.dart [hunk 5]
```diff
@@ -0,0 +1,154 @@
...
+      _onLiveStatus(LiveStatus.AGORA_JOIN_THE_ROOM);
+
+      _agoraStore.announceBroadcastReady(_roomDocument);
+
+      if (isRoomClosed) return dispose(false);
+      _onLiveStatus(LiveStatus.MASTER_READY);
+    }
+    else
+    {
+      _clientStore = ClientStore();
```
Comment (AlexVegner) [inline mobile/lib/modules/live/model/store/LiveStore.dart:122]: Please inject it, to be sure that we could write unit test in future.

File: mobile/lib/modules/live/model/store/RoomsStore.dart [hunk 6]
```diff
@@ -0,0 +1,101 @@
+import 'dart:async';
```
Comment (AlexVegner) [inline mobile/lib/modules/live/model/store/RoomsStore.dart:1]: By Dart convention file and folder names should be in snake case: `room_controller.dart`

File: mobile/lib/modules/live/model/store/RoomsStore.dart [hunk 7]
```diff
@@ -0,0 +1,101 @@
...
+import 'package:highvibe/modules/live/model/vo/RoomVO.dart';
+import 'package:mobx/mobx.dart';
+
+import '../vo/RoomVO.dart';
+
+part 'RoomsStore.g.dart';
+
+enum RoomsStatus { CONNECTING, EMPTY, READY, ERROR }
+
+class RoomsStore = _RoomsStore with _$RoomsStore;
```
Comment (AlexVegner) [inline mobile/lib/modules/live/model/store/RoomsStore.dart:15]: Rename it as RoomController

File: mobile/lib/modules/live/model/store/live/AgoraStore.dart [hunk 8]
```diff
@@ -0,0 +1,140 @@
+import 'dart:async';
```
Comment (AlexVegner) [inline mobile/lib/modules/live/model/store/live/AgoraStore.dart:1]: By Dart convention file and folder names should be in snake case: `agora_controller.dart`

File: mobile/lib/modules/live/model/store/live/AgoraStore.dart [hunk 9]
```diff
@@ -0,0 +1,140 @@
...
+import '../../vo/RoomVO.dart';
+import 'enum/LiveStatus.dart';
+import 'mixin/LiveStatusDispatcher.dart';
+
+// part 'AgoraStore.g.dart';
+//
+// class AgoraStore = _AgoraStore with _$AgoraStore;
+//
+// abstract class _AgoraStore with Store, LiveStatusDispatcher {
+class AgoraStore with LiveStatusDispatcher {
```
Comment (AlexVegner) [inline mobile/lib/modules/live/model/store/live/AgoraStore.dart:19]: Use class name AgoraController

File: mobile/lib/modules/live/model/store/live/AgoraStore.dart [hunk 10]
```diff
@@ -0,0 +1,140 @@
...
+    sendStatus(LiveStatus.AGORA_CREATED);
+  }
+
+  Future<void> initialize(RoomVO roomVO, Size size) async
+  {
+    print('> AgoraStore -> initialize: role = ${roomVO.role}');
+    print('> AgoraStore -> initialize: room = ${roomVO.appId}');
+    _engine = await RtcEngine.create(roomVO.appId);
+    await _engine.enableVideo();
+    if (_engine == null) return;
```
Comment (AlexVegner) [inline mobile/lib/modules/live/model/store/live/AgoraStore.dart:35]: Do we need 

```dart
 if (_engine == null) return;
```

each time when we use `_engine?

File: mobile/lib/modules/live/model/store/live/ClientStore.dart [hunk 11]
```diff
@@ -0,0 +1,55 @@
+import 'dart:async';
```
Comment (AlexVegner) [inline mobile/lib/modules/live/model/store/live/ClientStore.dart:1]: By Dart convention file and folder names should be in snake case: live_client_controller.dart

File: mobile/lib/modules/live/model/store/live/ClientStore.dart [hunk 12]
```diff
@@ -0,0 +1,55 @@
...
+import '../../vo/RoomVO.dart';
+import 'package:mobx/mobx.dart';
+import 'mixin/LiveStatusDispatcher.dart';
+import 'enum/LiveStatus.dart';
+
+import 'PeersStore.dart';
+
+part 'ClientStore.g.dart';
+
+class ClientStore = _ClientStore with _$ClientStore;
```
Comment (AlexVegner) [inline mobile/lib/modules/live/model/store/live/ClientStore.dart:13]: Rename it with LiveClientController

File: mobile/lib/modules/live/model/store/live/MasterStore.dart [hunk 13]
```diff
@@ -0,0 +1,68 @@
+import 'dart:async';
```
Comment (AlexVegner) [inline mobile/lib/modules/live/model/store/live/MasterStore.dart:1]: By Dart convention file and folder names should be in snake case: live_master_controller.dart

File: mobile/lib/modules/live/model/store/live/MasterStore.dart [hunk 14]
```diff
@@ -0,0 +1,68 @@
...
+import '../../vo/RoomVO.dart';
+import 'package:mobx/mobx.dart';
+import 'mixin/LiveStatusDispatcher.dart';
+import 'enum/LiveStatus.dart';
+
+import 'PeersStore.dart';
+
+part 'MasterStore.g.dart';
+
+class MasterStore = _MasterStore with _$MasterStore;
```
Comment (AlexVegner) [inline mobile/lib/modules/live/model/store/live/MasterStore.dart:13]: Rename it with LiveMasterController

File: mobile/lib/modules/live/model/store/live/mixin/LiveStatusDispatcher.dart [hunk 15]
```diff
@@ -0,0 +1,18 @@
+import 'dart:async';
```
Comment (AlexVegner) [inline mobile/lib/modules/live/model/store/live/mixin/LiveStatusDispatcher.dart:1]: By Dart convention file and folder names should be in snake case `live_status_dispatcher.dart`

File: mobile/lib/modules/live/model/vo/TokenVO.dart [hunk 16]
```diff
@@ -0,0 +1,18 @@
+class TokenVO {
```
Comment (AlexVegner) [inline mobile/lib/modules/live/model/vo/TokenVO.dart:1]: Please move it to core/lib/models/live/live_token.dart
Rename class LiveToken

File: mobile/lib/modules/live/components/pages/live.page.dart [hunk 17]
```diff
@@ -0,0 +1,71 @@
...
+              liveStore.isMaster
+              ? LiveMaster()
+              : LiveClient()
+            )
+          ),
+      );
+    });
+  }
+
+  Future<bool> _onBackPressed(context, LiveStore liveStore) async {
```
Comment (AlexVegner) [inline mobile/lib/modules/live/components/pages/live.page.dart:35]: User can click on Android hardware back button or use back gesture on iOS. 
We need to handle the case. 
Please handle it or write `//TODO` for it.

**Review by AlexVegner — CHANGES_REQUESTED** [review]
Multiple changes required
Also please check linter it show errors for your PR

File: mobile/lib/modules/live/model/store/live/enum/LiveStatus.dart [hunk 18]
```diff
@@ -0,0 +1,67 @@
+enum LiveStatus {
```
[the remark this answers is not in the document]
Comment (vladimircores) [reply mobile/lib/modules/live/model/store/live/enum/LiveStatus.dart]: Done.

File: mobile/pubspec.yaml [hunk 19]
```diff
@@ -64,7 +67,7 @@ dev_dependencies:
   built_value_generator: ^7.1.0
   built_value_test: ^7.1.0
   dependency_validator: ^1.0.0
-  mock_cloud_firestore: ^0.0.20
+  mock_cloud_firestore:
```
[the remark this answers is not in the document]
Comment (vladimircores) [reply mobile/pubspec.yaml:69]: Yeah, I forgot to revert it back, thanks for pointing it.

File: mobile/lib/modules/live/model/store/live/MasterStore.dart [hunk 20]
```diff
@@ -0,0 +1,68 @@
...
+          if (_peersStore.hasPeer(peerID)) {
+            _peersStore.unregisterAndClosePeer(peerID);
+            if (_isClosed) return dispose(false);
+            sendStatus(LiveStatus.MASTER_PEER_REMOVED);
+          }
+        }
+        else if (peerChange.type == DocumentChangeType.added)
+        {
+          if (!_peersStore.hasPeer(peerID)) {
+            // DocumentReference peerDocument = _peersStore.peersCollection.doc(
```
[the remark this answers is not in the document]
Comment (vladimircores) [reply mobile/lib/modules/live/model/store/live/MasterStore.dart]: It's not needed for now, and can be grabbed from original version. Removed.

File: mobile/lib/modules/live/components/pages/live/live.client.widget.dart [hunk 21]
```diff
@@ -0,0 +1,143 @@
...
+          if (_spinnerWidget != null) stackChildren.add(_spinnerWidget);
+
+          return Stack(children: stackChildren);
+        },
+        ),
+      );
+    });
+  }
+
+  // Future<void> _onClientDisposed(context) async {
```
[the remark this answers is not in the document]
Comment (vladimircores) [reply mobile/lib/modules/live/components/pages/live/live.client.widget.dart:88]: It's not needed for now, can be returned from original code if needed. Removed.

File: mobile/lib/modules/live/components/pages/live/live.master.widget.dart [hunk 22]
```diff
@@ -0,0 +1,127 @@
...
+Widget _spinnerWidget = _createSpinner();
+
+Key videoKey = const Key('VIDEO_UNIQUE_KEY');
+
+class LiveMaster extends StatelessWidget {
+  LiveMaster() {
+    _statusChildren = <Widget>[];
+  }
+
+  static const _tsMessage = TextStyle(color: Colors.red, fontSize: 14.0);
```
[the remark this answers is not in the document]
Comment (vladimircores) [reply mobile/lib/modules/live/components/pages/live/live.master.widget.dart]: Thanks for advice, I will in the future, these styles is temporal and will be removed.

File: mobile/lib/modules/live/components/pages/rooms.page.dart [hunk 23]
```diff
@@ -0,0 +1,104 @@
...
+import 'package:flutter_mobx/flutter_mobx.dart';
+import 'package:highvibe/modules/live/components/widgets/SelectableList.dart';
+import 'package:highvibe/modules/live/model/store/RoomsStore.dart';
+import 'package:provider/provider.dart';
+
+import '../../live_module.dart';
+
+class RoomsPage extends StatelessWidget {
+
+  static const COLOR_BACK = Colors.white;
```
[the remark this answers is not in the document]
Comment (vladimircores) [reply mobile/lib/modules/live/components/pages/rooms.page.dart]: Thanks for advice, I will in the future, these styles is temporal and will be removed.

File: mobile/lib/modules/live/components/pages/rooms.page.dart [hunk 24]
```diff
@@ -0,0 +1,104 @@
...
+  static const COLOR_TEXT = Colors.black;
+
+  static const _textStyle_Message
```
[the remark this answers is not in the document]
Comment (vladimircores) [reply mobile/lib/modules/live/components/pages/rooms.page.dart]: Thanks for advice, I will in the future, these styles is temporal and will be removed.

File: mobile/lib/modules/live/live_page.dart [hunk 25]
```diff
@@ -8,15 +8,15 @@ import 'package:highvibe/widgets/event_card.dart';
 import 'package:highvibe/widgets/header_row.dart';
 import 'package:highvibe/widgets/stream_card.dart';
 
-class LivePage extends StatefulWidget {
+class LivePage2 extends StatefulWidget {
```
[the remark this answers is not in the document]
Comment (vladimircores) [reply mobile/lib/modules/live/live_page.dart:12]: I would like to use it as a reference for RoomsPage implementation. Will keep it for now with postfix in name Template.

File: mobile/lib/modules/live/model/store/live/AgoraStore.dart (hunk 10 above)
Comment (vladimircores) [reply mobile/lib/modules/live/model/store/live/AgoraStore.dart]: It was made to avoid silent try/catch but now change to use it.

**Review by AlexVegner — APPROVED** [review]

**Pull request merged**